File: typescript/tests/p256-fixture.test.ts
/**
 * Cross-language P-256 ECDSA verify fixture runner (TypeScript).
 *
 * Drives Ed25519Service.verifyWithFallback through the shared corpus at
 * `tests/cross-language/p256-fixtures.json` — DER SubjectPublicKeyInfo public key +
 * ASN.1 DER ECDSA signature + SHA-256, per PROTOCOL_SPEC.md §7.5. Every AetherNet SDK
 * drives the SAME vectors and MUST accept valid:true and reject valid:false.
 *
 * Run with: tsx --test typescript/tests/p256-fixture.test.ts
 * SPDX-License-Identifier: MIT
 */
import { describe, it } from "node:test";
import { strict as assert } from "node:assert";
import { readFileSync, existsSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

import { Ed25519Service } from "../src/security/Ed25519Service.js";

interface Vector {
  name: string;
  public_key_der: string;
  message: string;
  signature_der: string;
  valid: boolean;
}
interface Corpus {
  vectors: Vector[];
}

function loadCorpus(): Corpus {
  let dir = dirname(fileURLToPath(import.meta.url));
  for (;;) {
    const candidate = join(dir, "tests", "cross-language", "p256-fixtures.json");
    if (existsSync(candidate)) {
      return JSON.parse(readFileSync(candidate, "utf8")) as Corpus;
    }
    const parent = dirname(dir);
    if (parent === dir) {
      throw new Error(
        "p256-fixtures.json not found walking up from " +
          dirname(fileURLToPath(import.meta.url)),
      );
    }
    dir = parent;
  }
}

describe("P-256 ECDSA verify — cross-language fixture", () => {
  const corpus = loadCorpus();
  assert.ok(corpus.vectors.length > 0, "no vectors");

  for (const v of corpus.vectors) {
    it(v.name, () => {
      const pub = Buffer.from(v.public_key_der, "hex");
      const msg = Buffer.from(v.message, "hex");
      const sig = Buffer.from(v.signature_der, "hex");
      // A >32-byte key forces the P-256 branch; a regression to the old
      // "return false" stub would reject the valid vector and fail here.
      assert.ok(pub.length > 32, "P-256 SPKI key must be > 32 bytes");
      assert.equal(Ed25519Service.verifyWithFallback(pub, msg, sig), v.valid);
    });
  }
});

File: go/security/p256_fixture_test.go
// SPDX-License-Identifier: MIT

package security

import (
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

// TestVerifyWithFallback_P256Fixture drives Ed25519Service.VerifyWithFallback through
// the shared cross-language corpus at tests/cross-language/p256-fixtures.json — DER
// SubjectPublicKeyInfo public key + ASN.1 DER ECDSA signature + SHA-256, per
// PROTOCOL_SPEC.md §7.5. Every AetherNet SDK drives the SAME vectors and MUST accept
// valid:true and reject valid:false.
func TestVerifyWithFallback_P256Fixture(t *testing.T) {
	raw, err := os.ReadFile(findP256Fixture(t))
	if err != nil {
		t.Fatal(err)
	}
	var doc struct {
		Vectors []struct {
			Name         string `json:"name"`
			PublicKeyDer string `json:"public_key_der"`
			Message      string `json:"message"`
			SignatureDer string `json:"signature_der"`
			Valid        bool   `json:"valid"`
		} `json:"vectors"`
	}
	if err := json.Unmarshal(raw, &doc); err != nil {
		t.Fatal(err)
	}
	if len(doc.Vectors) == 0 {
		t.Fatal("no vectors in p256-fixtures.json")
	}

	es := NewEd25519Service()
	for _, v := range doc.Vectors {
		pub, _ := hex.DecodeString(v.PublicKeyDer)
		msg, _ := hex.DecodeString(v.Message)
		sig, _ := hex.DecodeString(v.SignatureDer)
		// A >32-byte key forces the P-256 branch; a regression to Ed25519-only would
		// reject the valid vector and fail here.
		if len(pub) <= 32 {
			t.Fatalf("%s: P-256 SPKI key must be > 32 bytes", v.Name)
		}
		if got := es.VerifyWithFallback(pub, msg, sig); got != v.Valid {
			t.Errorf("%s: VerifyWithFallback = %v, want %v", v.Name, got, v.Valid)
		}
	}
}

func findP256Fixture(t *testing.T) string {
	dir, err := os.Getwd()
	if err != nil {
		t.Fatal(err)
	}
	for {
		cand := filepath.Join(dir, "tests", "cross-language", "p256-fixtures.json")
		if _, statErr := os.Stat(cand); statErr == nil {
			return cand
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			t.Fatalf("p256-fixtures.json not found walking up from working dir")
		}
		dir = parent
	}
}
